src/main/groovy/com/rundeck/plugins/ansible/plugin/AnsibleResourceModelSource.java

```diff
@@ -345,47 +371,56 @@ public AnsibleRunner.AnsibleRunnerBuilder buildAnsibleRunner() throws ResourceMo
     return runnerBuilder;
   }
 
-
   @Override
   public INodeSet getNodes() throws ResourceModelSourceException {
     NodeSetImpl nodes = new NodeSetImpl();
-    final Gson gson = new Gson();
+    AnsibleRunner.AnsibleRunnerBuilder runnerBuilder = buildAnsibleRunner();
 
+    if (gatherFacts) {
+      processWithGatherFacts(nodes, runnerBuilder);
+    } else {
+      ansibleInventoryList(nodes, runnerBuilder);
+    }
+
+    return nodes;
+  }
+
+  public void processWithGatherFacts(NodeSetImpl nodes, AnsibleRunner.AnsibleRunnerBuilder runnerBuilder) throws ResourceModelSourceException {
+
+    final Gson gson = new Gson();
     Path tempDirectory;
     try {
       tempDirectory = Files.createTempDirectory("ansible-hosts");
     } catch (IOException e) {
-        throw new ResourceModelSourceException("Error creating temporary directory.", e);
+      throw new ResourceModelSourceException("Error creating temporary directory.", e);
     }
 
     try {
-      Files.copy(this.getClass().getClassLoader().getResourceAsStream("host-tpl.j2"), tempDirectory.resolve("host-tpl.j2"));
-      Files.copy(this.getClass().getClassLoader().getResourceAsStream("gather-hosts.yml"), tempDirectory.resolve("gather-hosts.yml"));
+      Files.copy(this.getClass().getClassLoader().getResourceAsStream(HOST_TPL_J2), tempDirectory.resolve(HOST_TPL_J2));
+      Files.copy(this.getClass().getClassLoader().getResourceAsStream(GATHER_HOSTS_YML), tempDirectory.resolve(GATHER_HOSTS_YML));
     } catch (IOException e) {
-        throw new ResourceModelSourceException("Error copying files.");
+      throw new ResourceModelSourceException("Error copying files.");
     }
 
-    AnsibleRunner.AnsibleRunnerBuilder runnerBuilder = buildAnsibleRunner();
-
     runnerBuilder.tempDirectory(tempDirectory);
     runnerBuilder.retainTempDirectory(true);
 
     StringBuilder args = new StringBuilder();
     args.append("facts: ")
-        .append(gatherFacts ? "True" : "False")
-        .append("\n")
-        .append("tmpdir: '")
-        .append(tempDirectory.toFile().getAbsolutePath())
-        .append("'");
+            .append(gatherFacts ? "True" : "False")
+            .append("\n")
+            .append("tmpdir: '")
+            .append(tempDirectory.toFile().getAbsolutePath())
+            .append("'");
 
     runnerBuilder.extraVars(args.toString());
 
     AnsibleRunner runner = runnerBuilder.build();
 
     try {
-        runner.run();
+      runner.run();
     } catch (Exception e) {
-        throw new ResourceModelSourceException(e.getMessage(),e);
+      throw new ResourceModelSourceException(e.getMessage(),e);
```

Comment:
Here i would prefer to, if possible, add a custom message with valuable information for the user. If not possible to have a custom message that can provide this kind of info we could just set the exception without the message, this would propagate the exception and it should only change the exception type, assuming that rundeck handles this kind of exception (ResourceModelSourceException) on a different way than just a regular exception
```suggestion
      throw new ResourceModelSourceException("CUSTOM MESSAGE with valuable information",e);
```

Reply:
Done... ✅